### PR TITLE
feat: make greatest supports timestamp and datetime types

### LINKED ISSUE #4991: Add support for all comparable types for `greatest` function
### What problem does the new feature solve?

The `greatest` function only supports `date`. It would be great if it could support all comparable types. For now I have to use  `CASE WHEN` to get the greatest value.

### What does the feature do?

support `select greatest(v1, 100)` etc. 

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by killme2008**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

Close #4991 

## What's changed and what's your intention?

Enhance the `greatest(a, b)` function to:

* Accept `DateTime` and `Timestamp` inputs.
* Return `DateTime` for string inputs instead of `Date`.

If DataFusion supports `greatest` and `least` in the future, we can remove this function.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/common/function/src/scalars/timestamp/greatest.rs [hunk 1]
```diff
@@ -87,6 +117,47 @@ impl Function for GreatestFunction {
...
+                let result =
+                    zip::zip(&boolean_array, &column1, &column2).context(ArrowComputeSnafu)?;
+                Ok(Helper::try_into_vector(&result).context(error::FromArrowArraySnafu)?)
+            }
+            ConcreteDataType::Timestamp(ts_type) => {
+                let column1 = columns[0].to_arrow_array();
+                let column2 = columns[1].to_arrow_array();
+
+                let boolean_array = match ts_type {
+                    TimestampType::Second(_) => {
```
Comment (v0y4g3r) [inline src/common/function/src/scalars/timestamp/greatest.rs:135]: Can we use macro to simplify thos match arms?

File: src/common/function/src/scalars/timestamp/greatest.rs [hunk 2]
```diff
@@ -87,6 +117,47 @@ impl Function for GreatestFunction {
...
+                    TimestampType::Nanosecond(_) => {
+                        let column1 = column1.as_primitive::<TimestampNanosecondType>();
+                        let column2 = column2.as_primitive::<TimestampNanosecondType>();
+                        gt(&column1, &column2).context(ArrowComputeSnafu)?
+                    }
+                };
+
+                let result =
+                    zip::zip(&boolean_array, &column1, &column2).context(ArrowComputeSnafu)?;
+                Ok(Helper::try_into_vector(&result).context(error::FromArrowArraySnafu)?)
```
Comment (v0y4g3r) [inline src/common/function/src/scalars/timestamp/greatest.rs:159]:
```suggestion
                Helper::try_into_vector(&result).context(error::FromArrowArraySnafu)
```

File: src/common/function/src/scalars/timestamp/greatest.rs (hunk 1 above)
Comment (killme2008) [reply src/common/function/src/scalars/timestamp/greatest.rs]: refactor at https://github.com/GreptimeTeam/greptimedb/pull/5005/commits/657e992a5a73282e31f4a944170f8aa575bf908c

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]

**Pull request merged**